crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -723,7 +736,83 @@ impl UnableToDecryptReason {
     /// Returns true if this UTD is due to a missing room key (and hence might
     /// resolve itself if we wait a bit.)
     pub fn is_missing_room_key(&self) -> bool {
-        matches!(self, Self::MissingMegolmSession | Self::UnknownMegolmMessageIndex)
+        // In case of MissingMegolmSession with a withheld code we return false here
+        // given that this API is used to decide if waiting a bit will help.
+        matches!(self, Self::MissingMegolmSession(None) | Self::UnknownMegolmMessageIndex)
+    }
+}
+
+/// A machine-readable code for why the megolm key was not sent.
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    AsStrAsRefStr,
+    AsRefStr,
+    FromString,
+    DebugAsRefStr,
+    SerializeAsRefStr,
+    DeserializeFromCowStr,
+)]
+#[non_exhaustive]
+pub enum WithheldCode {
+    /// the user/device was blacklisted.
+    #[ruma_enum(rename = "m.blacklisted")]
+    Blacklisted,
+
+    /// the user/devices is unverified.
+    #[ruma_enum(rename = "m.unverified")]
+    Unverified,
+
+    /// The user/device is not allowed have the key. For example, this would
+    /// usually be sent in response to a key request if the user was not in
+    /// the room when the message was sent.
+    #[ruma_enum(rename = "m.unauthorised")]
+    Unauthorised,
+
+    /// Sent in reply to a key request if the device that the key is requested
+    /// from does not have the requested key.
+    #[ruma_enum(rename = "m.unavailable")]
+    Unavailable,
+
+    /// An olm session could not be established.
+    /// This may happen, for example, if the sender was unable to obtain a
+    /// one-time key from the recipient.
+    #[ruma_enum(rename = "m.no_olm")]
+    NoOlm,
+
+    #[doc(hidden)]
+    _Custom(PrivOwnedStr),
+}
+
+impl fmt::Display for WithheldCode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let string = match self {
+            WithheldCode::Blacklisted => "The sender has blocked you.",
+            WithheldCode::Unverified => "The sender has disabled encrypting to unverified devices.",
+            WithheldCode::Unauthorised => "You are not authorised to read the message.",
+            WithheldCode::Unavailable => "The requested key was not found.",
+            WithheldCode::NoOlm => "Unable to establish a secure channel.",
+            _ => self.as_str(),
+        };
+
+        f.write_str(string)
+    }
+}
+
+// Wrapper around `Box<str>` that cannot be used in a meaningful way outside of
+// this crate. Used for string enums because their `_Custom` variant can't be
+// truly private (only `#[doc(hidden)]`).
+#[doc(hidden)]
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+// XXX had to make the field public to make ruma `StringEnum` macro happy.
+pub struct PrivOwnedStr(pub Box<str>);
```

Comment:
Yes nothing makes sense with the struct anymore (the name, the doc, ..)
Not sure how to properly document it though.

Reply:
The Ruma macro expects the type to have this name, it's not something that is in under our control...

The alternatives are to not use the Ruma macro or to have multiple copies of this struct.

Reply:
ok, fair enough. It would be helpful if this could be explained in a comment.

Why *does* this need to be `pub`, when it didn't need to be `pub` when it lived in the crypto crate?

Reply:
https://github.com/matrix-org/matrix-rust-sdk/pull/4305/commits/1f50e4b1a04b9b3f2f408a9a5b843459e97d711a

> when it didn't need to be pub when it lived in the crypto crate

Because it is also used by other enums (still in crypto) that now uses the one that is in common now.

Reply:
Actually there was another copy left https://github.com/matrix-org/matrix-rust-sdk/pull/4305/commits/2151fcdbe84c09c051a9c42ae9761620c63e1822

I would be ok to revert that and have multiple copies if it makes it simpler

Reply:
> Why does this need to be pub, when it didn't need to be pub when it lived in the crypto crate?

Because otherwise you can't construct the variant outside of the crate where the type lives unless it's pub. Now I don't exactly know where that is needed but if we remove the pub the compiler might tell us.